### PR TITLE
Upload, upload tests, fix rate with git install in docker containers, rate tests

### PR DESCRIPTION
**Pull request by priyankapereira1**
Completed package upload with tests.

Upload has 2 options, ingestible public URL input, or base64 encoded zip input.
Stores packages that are inputted and/or git cloned in /tmp/npm-package-**** just like how it was for your Part 1 for simpler integration.

This WILL BE CHANGED! Simple fix on next iteration.

Tests in tests/integration_tests/package_upload_and_rate.test.ts do not seed the database, but run upload on various inputs to populate the database, then for one successful rating check on one of the uploaded packages (need a valid cloned/stored location for rating endpoint to work).

Tests in tests/integration_tests/package_rate.test.ts only do invalid checks on the rating endpoint.

### CODE REVIEW

**Comment by priyankapereira1** [comment]
Will fix the FK_UserID fields when creating a database entry on upload next, on a new branch. Feel free to look over and merge if ready.

File: src/api_server/zip_files.ts
Comment (just007in) [inline src/api_server/zip_files.ts:1]: Please use writeFile from 'fs/promises' package. Line 67

File: src/api_server/get_files.ts
Comment (just007in) [inline src/api_server/get_files.ts:1]: Please use readFile from 'fs/promises'

File: src/api_server/db_connector.ts
Comment (just007in) [inline src/api_server/db_connector.ts:1]: Is there a reason to change packageid to be non unique string?

File: docker/API/Dockerfile
Comment (just007in) [inline docker/API/Dockerfile:1]: We should instead use isomorphic git package if possible

**Review by just007in — CHANGES_REQUESTED** [review]

File: src/api_server/db_connector.ts
Comment (priyankapereira1) [reply src/api_server/db_connector.ts:1]: example response had a lowercase string as PackageID
models also has PackageID as string

File: src/api_server/get_files.ts
Comment (priyankapereira1) [reply src/api_server/get_files.ts:1]: can do, but what is the rationale?

File: src/api_server/zip_files.ts
Comment (priyankapereira1) [reply src/api_server/zip_files.ts:1]: rationale?

File: docker/API/Dockerfile
Comment (priyankapereira1) [reply docker/API/Dockerfile:1]: this should work, is there a reason it would not?

File: src/api_server/zip_files.ts
Comment (just007in) [reply src/api_server/zip_files.ts:1]: Synchronous functions blocks the whole main thread so using the async function instead is better

File: src/api_server/get_files.ts
Comment (just007in) [reply src/api_server/get_files.ts:1]: Synchronous functions blocks the whole main thread so using the async function instead is better

File: src/api_server/db_connector.ts
Comment (just007in) [reply src/api_server/db_connector.ts:1]: Ah OK didn't realize that

File: docker/API/Dockerfile
Comment (just007in) [reply docker/API/Dockerfile:1]: Yeah this should work. I thought we were going to use isomorphic git

File: src/api_server/zip_files.ts
Comment (priyankapereira1) [reply src/api_server/zip_files.ts:1]: on it! newest commit should fix this

File: src/api_server/get_files.ts
Comment (priyankapereira1) [reply src/api_server/get_files.ts:1]: on it, newest commit should fix this :)

**Comment by priyankapereira1** [comment]
> Will fix the FK_UserID fields when creating a database entry on upload next, on a new branch. Feel free to look over and merge if ready.

fixed on newest commit

**Review by just007in — APPROVED** [review]

**Pull request merged**